pkg/watermark/fetch/edge_fetcher.go

```diff
@@ -202,14 +207,22 @@ func (e *edgeFetcher) GetHeadWMB(fromPartitionIdx int32) wmb.WMB {
 		return wmb.WMB{}
 	}
 
-	e.Lock()
-	defer e.Unlock()
 	// update the last processed watermark for the partition
+	e.Lock()
 	e.lastProcessedWm[fromPartitionIdx] = headWMB.Watermark
+	e.Unlock()
+
```

Comment:
I like putting a lock around the smallest section of code that requires it

Reply:
how about using https://pkg.go.dev/sync#RWMutex

Reply:
that would be more optimal